### PR TITLE
Make resetting on experiment change optional

### PR DESCRIPTION
**Pull request by moggyboy**
We would prefer never to reset the data even if the variants change names, the metadata changes, etc. I've added a `reset_manually` configuration flag that makes this possible, and refactored the Redis persistence so that it applies incremental changes rather than blowing away all experiment data before it's rewritten.

### CODE REVIEW

**Comment by moggyboy** [comment]
I'm very confused as to why this is failing - all specs pass locally?

**Comment by moggyboy** [comment]
Hello Mr/Ms Reviewer - I recommend a commit-by-commit review as I took the liberty of cleaning up a bunch of code in the Experiment class, which will obfuscate the actual change being made.

**Comment by andrew** [comment]
@moggyboy would you be able to split the cleanup/refactoring changes into a separate PR so it's a little easier for me to review them?

**Comment by moggyboy** [comment]
@andrew All but the last commit are necessary refactorings in order to implement the feature. If you step through them commit-by-commit, it shouldn't be too hard to follow?

**Comment by moggyboy** [comment]
@andrew Is there anything else I can do to help this get over the line?

**Comment by andrew** [comment]
@moggyboy if you can pop that last commit off I can get it merged

**Comment by moggyboy** [comment]
Done @andrew :-)

**Comment by andrew** [comment]
Thanks!

**Comment by zinkkrysty** [comment]
I discovered this setting by browsing the code. Should there not be a documentation in the readme about it?

**Comment by andrew** [comment]
@zinkkrysty that sounds like a good idea, would you be up for contributing? Or perhaps opening a fresh issue for this so it doesn't get lost

**Comment by zinkkrysty** [comment]
@andrew sure thing: #534
